[rustdoc] Wrap code blocks in <code> tag

Kobzol opened the pull request:
This PR modifies Rustdoc output so that fenced code snippets, items and whole file source codes are wrapped in `<pre><code>` instead of just `<pre>`. This should improve the semantic meaning of the generated content.

I'm not sure what to do about `render_attributes_in_pre` and `render_attributes_in_code`. These functions were clearly expected to be used for things inside `<pre>` or `<code>`, and since I added `<code>` in this PR, some of them will be used in a different context than before. However, it seems to me that even before they were not consistent. For example, `item_constant` used `render_attributes_in_code` for its attributes, however there was no `<code>` used for constants before this PR...

Should I create some `rustdoc-gui` tests? For example to check that all `<pre>` tags have a `<code>` child?

Fixes: https://github.com/rust-lang/rust/issues/88020

rust-highfive commented:
r? @ollie27

(rust-highfive has picked a reviewer for you, use r? to override)

GuillaumeGomez commented on src/librustdoc/html/highlight.rs:
```diff
@@ -69,6 +69,7 @@ fn write_header(out: &mut Buffer, class: Option<&str>, extra_content: Option<Buf
     } else {
         writeln!(out, "<pre class=\"rust\">");
     }
+    write!(out, "<code>");
```
I think this is not ideal because before we have `writeln!`, what about:

```rust
    if let Some(class) = class {
        writeln!(out, "<pre class=\"rust {}\"><code>", class);
    } else {
        writeln!(out, "<pre class=\"rust\"><code>");
    }
```

?

Kobzol replied:
I did it in this way because if I also used `writeln!` here, it would generate `<pre>\n<code>\n`, which added a new empty line at the beginning of the code block. I originally wanted to use the code that you have posted, but it seemed like a needless duplication of `<code>`.

Is there something that you don't like with the current code here? I'm not sure what is the problem.

GuillaumeGomez replied:
I want the backline after the tags. If the backline isn't needed, then better to replace `writeln!` with `write!`.

Kobzol replied:
I see. I think that the newline isn't actually even needed, so I'll try to remove it. If it's needed, I'll put it after the tags.

GuillaumeGomez commented on src/librustdoc/html/render/print_item.rs:
```diff
@@ -1322,6 +1341,15 @@ where
     w.write_str("</div>")
 }
 
+fn wrap_item<F>(w: &mut Buffer, item_name: &str, f: F)
```
Great idea! :+1:

Kobzol replied:
I simply borrowed it from `wrap_into_docblock` above :)

GuillaumeGomez commented on src/test/rustdoc-gui/type-declation-overflow.goml:
```diff
@@ -5,4 +5,4 @@ size: (1100, 800)
 assert-property: ("body", {"scrollWidth": "1100"})
 // However, since there is overflow in the type declaration, its scroll width is bigger.
-assert-property: (".type-decl pre", {"scrollWidth": "1324"})
+assert-property: (".type-decl pre", {"scrollWidth": "1332"})
```
Normally, this shouldn't have changed...

Kobzol replied:
I reverted the change to the test and instead changed CSS to remove padding from `<pre><code>` blocks.

GuillaumeGomez commented:
> I'm not sure what to do about `render_attributes_in_pre` and `render_attributes_in_code`. These functions were clearly expected to be used for things inside `<pre>` or `<code>`, and since I added `<code>` in this PR, some of them will be used in a different context than before. However, it seems to me that even before they were not consistent. For example, `item_constant` used `render_attributes_in_code` for its attributes, however there was no `<code>` used for constants before this PR...

Another shortcoming of rustdoc... Welcome to a huge years old codebase! :)

For consistency it should be applied the same on equivalent items but I think it's a bit out of scope here. For things other than codeblocks, just having a `<code>` (without the `<pre>` wrapping) is fine.

> Should I create some `rustdoc-gui` tests? For example to check that all `<pre>` tags have a `<code>` child?

Well, not all `<code>` are supposed to be wrapped. For example, an inline code in documentation like `this one` isn't supposed to be wrapped into a `<pre>`.

jyn514 commented:
r? @GuillaumeGomez

Kobzol commented:
I changed the newlines after `<pre><code>`, now no newline is generated. I also added a simple test that checks that a few item types and doc code blocks contain `<pre><code>`.

GuillaumeGomez commented on src/librustdoc/html/render/print_item.rs:
```diff
@@ -455,24 +455,25 @@ fn item_function(w: &mut Buffer, cx: &Context<'_>, it: &clean::Item, f: &clean::
+        write!(
+            w,
+            "{vis}{constness}{asyncness}{unsafety}{abi}fn \
+         {name}{generics}{decl}{notable_traits}{where_clause}",
```
Please keep the alignment in the string as well (nitpicking hello).

Kobzol replied:
I reverted the alignment as it was before. Thanks :)

GuillaumeGomez commented:
I checked locally and it looks good to me.

Thanks!

@bors: r+

bors commented:
:pushpin: Commit ccd550ee565436b4c9b7f987c19bd3a9111714f1 has been approved by `GuillaumeGomez`

<!-- @bors r=GuillaumeGomez ccd550ee565436b4c9b7f987c19bd3a9111714f1 -->
<!-- homu: {"type":"Approved","sha":"ccd550ee565436b4c9b7f987c19bd3a9111714f1","approver":"GuillaumeGomez"} -->
